text/3529-cargo-path-bases.md

```diff
@@ -0,0 +1,426 @@
+- Feature Name: `path_bases`
+- Start Date: 2023-11-13
+- RFC PR: [rust-lang/rfcs#3529](https://github.com/rust-lang/rfcs/pull/3529)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Introduce shared base directories in Cargo configuration files that in
+turn enable base-relative `path` dependencies.
+
+# Motivation
+[motivation]: #motivation
+
+As a project grows in size, it becomes necessary to split it into smaller
+sub-projects, architected into layers with well-defined boundaries.
+
+One way to enforce these boundaries is to use different Git repos (aka
+"multi-repo"). Cargo has good support for multi-repo projects using either `git`
+dependencies, or developers can use private registries if they want to
+explicitly publish code or need to preprocess their sub-projects (e.g.,
+generating code) before they can be consumed.
+
+If all of the code is kept in a single Git repo (aka "mono-repo"), then these
+boundaries must be enforced a different way: either leveraging tooling during
+the build to check layering, or requiring that sub-projects explicitly publish
+and consume from some intermediate directory. Cargo has poor support for
+mono-repos: the only viable mechanism is `path` dependencies, but these require

```

Comment:
I would assume the biggest issue for monorepos is scaling the builds.  If you use a registry, features like https://github.com/rust-lang/cargo/issues/5931 could help.  We have no plans in site for any kind of monorepo caching scheme.  This makes me wonder how viable Cargo is more generally for monorepos which makes me wonder how important this feature is.

Reply:
I'm not 100% sure what you mean by "scaling builds".

We certainly using caching for our monorepos, for example in the Office and Windows repos we use [BuildXL](https://github.com/microsoft/BuildXL) and model Cargo as a "pip" (i.e., a single unit of execution whose output is deterministic based on its input and so can be pulled from a cache). Eventually, we'd like to move to a model where each `rustc` invocation is a pip, but that requires some mechanism to query Cargo about what it plans to do (something like the deprecated "build plan" feature).

Nonetheless, this feature is required for both Windows and Office: we split our monorepos into "layers" which can only depend on artifacts "published" by a previous layer. Since the publishing directory is host and build-flavor dependent using `path` dependencies as they currently exist doesn't work, but the path base feature solves this problem since we can emit the absolute path to the publish directory in a generated config.toml when initializing the build.

Currently the "artifacts" that we are publishing and sharing is the source for crates (e.g., generated bindings for internal headers), but eventually we're going to need to use more dynamic libraries in our build graph. This means that the "artifacts" become a set of metadata about the functions exposed at the library boundary: akin to headers+import library or a .NET reference assembly. While I'm not sure what form this metadata will take (generated .rs files? .rmeta files? something new?), we'll still need some way to point Cargo at the "publish" directory where those files are located, so this feature is still relevant.

Reply:
> I'm not 100% sure what you mean by "scaling builds".

Build times

So you have a custom solution to help you with this problem and this RFC addressing another layer of monorepo onion of problems for you.  It feels a little strange that we are addressing a lower priority problem for monorepos when everyone has to reinvent solutions to the bigger problems, which would be prohibitive enough that I'm assuming MS will be the main monorepo user of this feature.  The "single user" aspect of this feature worries me.

For example, we have a lot of problems with pre-release versions but we identified what the biggest blocker is and are working on addressing that, setting aside work people have done for specific niches of pre-release they have.

However, in talking through this RFC, there has been a limited set of other users expressing interest in it that maybe we have enough of case for moving this forward; just not solely motivated by monorepos.